### PR TITLE
[RISC-V][LoongArch64] Pass FP struct fields at arbitrary offsets in ArgIterator and CallDescrWorker

### PR DESCRIPTION
**Pull request by tomeksowi**
Use new info calculated in #103945 on the VM side (shuffling thunks will be handled in a separate PR).

Stems from #101796, part of #84834, cc @dotnet/samsung @shushanhf @LuckyXu-HF

### CODE REVIEW

File: src/coreclr/vm/loongarch64/calldescrworkerloongarch64.S
Comment (tomeksowi) [inline src/coreclr/vm/loongarch64/calldescrworkerloongarch64.S:1]: @shushanhf @LuckyXu-HF PTAL, I don't really know LA asm and I had no way to test it..

**Comment by tomeksowi** [comment]
@MichalStrehovsky @jkotas can anyone review please?

**Comment by jkotas** [comment]
@dotnet/samsung @shushanhf @LuckyXu-HF Could you please review and sign-off?

**Comment by risc-vv** [comment]
<details><summary>8d8e705dis being scheduled for building and testing</summary>

GIT:: `8d8e705de3908f6138e7d691563776f8dc4e0ed5`
REPO: [dotnet/runtime](https://github.com/dotnet/runtime)
BRANCH: main</details>

**Comment by risc-vv** [comment]
<details><summary>8d8e705dis being scheduled for building and testing</summary>

GIT: `8d8e705de3908f6138e7d691563776f8dc4e0ed5`
REPO: [dotnet/runtime](https://github.com/dotnet/runtime)
BRANCH: main</details>

**Comment by risc-vv** [comment]
<details><summary>8d8e705d is being scheduled for building and testing</summary>

GIT: `8d8e705de3908f6138e7d691563776f8dc4e0ed5`
REPO: [dotnet/runtime](https://github.com/dotnet/runtime)
BRANCH: main</details>

**Review by sirntar — APPROVED** [review]
I left review of the LA64 parts for @shushanhf  and/or @LuckyXu-HF, but as for the RV64, everything looks good to me.

@tomeksowi I saw that the EmptyStructs test fails, but I'm not sure if this pr was supposed to solve this problem (?).

**Comment by shushanhf** [comment]
OK, Thanks very much.
I will test it based on the latest patch right now.

File: src/coreclr/vm/callingconvention.h [hunk 1]
```diff
@@ -909,6 +928,11 @@ class ArgIteratorTemplate : public ARGITERATOR_BASE
 protected:
     DWORD               m_dwFlags;              // Cached flags
     int                 m_nSizeOfArgStack;      // Cached value of SizeOfArgStack
+#if defined(TARGET_RISCV64) || defined(TARGET_RISCV64)
```
Comment (shushanhf) [inline src/coreclr/vm/callingconvention.h:931]:
```suggestion
#if defined(TARGET_RISCV64) || defined(TARGET_LOONGARCH64)
```

**Review by shushanhf — CHANGES_REQUESTED** [review]

File: src/coreclr/vm/loongarch64/calldescrworkerloongarch64.S [hunk 2]
```diff
@@ -76,127 +76,28 @@ LOCAL_LABEL(CallDescrWorkerInternalReturnAddress):
...
-    beq  $t4, $a3, LOCAL_LABEL(FloatIntReturn)
+    andi $a3, $a3, FpStruct__BothFloat
+    bne  $a3, $zero, LOCAL_LABEL(SecondFieldFloatReturn)
 
-    ori  $t4, $zero, CallDescrData__flagDoubleInt
-    beq  $t4, $a3, LOCAL_LABEL(DoubleIntReturn)
+    // The second returned register is integer (FpStruct::FloatInt | FpStruct::IntFloat)
+    // Note: it will also go in here for FpStruct::OnlyOne but storing a register of trash doesn't hurt
+    st.d $a0, $s0, CallDescrData__returnValue + 8
+    j  LOCAL_LABEL(ReturnDone)
```
Comment (shushanhf) [inline src/coreclr/vm/loongarch64/calldescrworkerloongarch64.S:93]:
```suggestion
    b  LOCAL_LABEL(ReturnDone)
```

File: src/coreclr/vm/loongarch64/calldescrworkerloongarch64.S [hunk 3]
```diff
@@ -76,127 +76,28 @@ LOCAL_LABEL(CallDescrWorkerInternalReturnAddress):
...
-    fst.d  $f1, $s0, CallDescrData__returnValue + 8
-    b  LOCAL_LABEL(ReturnDone)
-
-LOCAL_LABEL(DoubleDoubleReturn):
-    fst.d  $f0, $s0, CallDescrData__returnValue
-    fst.d  $f1, $s0, CallDescrData__returnValue + 8
-    b  LOCAL_LABEL(ReturnDone)
+LOCAL_LABEL(SecondFieldFloatReturn):
+    fst.d $fa1, $s0, CallDescrData__returnValue + 8
+    j  LOCAL_LABEL(ReturnDone)
```
Comment (shushanhf) [inline src/coreclr/vm/loongarch64/calldescrworkerloongarch64.S:97]:
```suggestion
    b  LOCAL_LABEL(ReturnDone)
```

**Comment by risc-vv** [comment]
<details><summary>3985f11e is being scheduled for building and testing</summary>

GIT: `3985f11e825beafbc9fd591d97942c4d69753333`
REPO: [dotnet/runtime](https://github.com/dotnet/runtime)
BRANCH: main</details>

**Review by shushanhf — APPROVED** [review]
LGTM
Thanks

**Comment by tomeksowi** [comment]
> @tomeksowi I saw that the EmptyStructs test fails, but I'm not sure if this pr was supposed to solve this problem (?).

This no, but #106112 disables them where not supported.

**Comment by risc-vv** [comment]
<details><summary>3985f11e is being scheduled for building and testing</summary>

GIT: `3985f11e825beafbc9fd591d97942c4d69753333`
REPO: [dotnet/runtime](https://github.com/dotnet/runtime)
BRANCH: main</details>

**Comment by risc-vv** [comment]
<details><summary>3985f11e is being scheduled for building and testing</summary>

GIT: `3985f11e825beafbc9fd591d97942c4d69753333`
REPO: [dotnet/runtime](https://github.com/dotnet/runtime)
BRANCH: main</details>

**Comment by risc-vv** [comment]
<details><summary>88815d2d is being scheduled for building and testing</summary>

GIT: `88815d2db74d1ac9039fe09874b81af9cdd9e47b`
REPO: [dotnet/runtime](https://github.com/dotnet/runtime)
BRANCH: main</details>

**Review by jkotas — APPROVED** [review]

**Pull request merged**